service/interceptor/interceptor.go

```diff
@@ -0,0 +1,232 @@
+		volNameLocks: map[string]gosync.TryLocker{},
+	}
+
+	gocsiSerializer := serialvolume.New(serialvolume.WithLockProvider(locker))
+
+	i := &interceptor{opts{locker: locker, timeout: 0}}
+
+	i.createMetadataRetrieverClient(context.Background())
+
+	handle := func(ctx xctx.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		switch t := req.(type) {
+		case *csi.CreateVolumeRequest:
+			return i.createVolume(ctx, t, info, handler)
+		case *csi.NodeStageVolumeRequest:
+			return i.nodeStageVolume(ctx, t, info, handler)
+		case *csi.NodeUnstageVolumeRequest:
+			return i.nodeUnstageVolume(ctx, t, info, handler)
+		default:
+			return gocsiSerializer(ctx, req, info, handler)
+		}
+	}
+	return handle
+}
+
+func (i *interceptor) createMetadataRetrieverClient(ctx context.Context) {
+	metricsManager := metrics.NewCSIMetricsManagerWithOptions("csi-metadata-retriever",
+		metrics.WithProcessStartTime(false),
+		metrics.WithSubsystem(metrics.SubsystemSidecar))
+
+	if retrieverAddress, ok := csictx.LookupEnv(ctx, "CSI_RETRIEVER_ENDPOINT"); ok {
+		rpcConn, err := connection.Connect(retrieverAddress, metricsManager, connection.OnConnectionLoss(connection.ExitOnConnectionLoss()))
+		if err != nil {
+			log.Error(err.Error())
+		}
+
+		retrieverClient := retriever.NewMetadataRetrieverClient(rpcConn, 100*time.Second)
+		if retrieverClient == nil {
+			log.Error("Cannot get csi-metadata-retriever client")
+		}
+
+		i.opts.MetadataSidecarClient = retrieverClient
+	} else {
+		log.Warn("env var not found: ", "CSI_RETRIEVER_ENDPOINT")
+	}
+}
+
+const pending = "pending"
+
+func (i *interceptor) nodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest,
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, resErr error) {
+	lock, err := i.opts.locker.GetLockWithID(ctx, req.VolumeId)
+	if err != nil {
+		return nil, err
+	}
+
+	if closer, ok := lock.(io.Closer); ok {
+		defer closer.Close() // #nosec G307
+	}
+
+	if !lock.TryLock(i.opts.timeout) {
+		return nil, status.Error(codes.Aborted, pending)
+	}
+	defer lock.Unlock()
+
+	return handler(ctx, req)
+}
+
+func (i *interceptor) nodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest,
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, resErr error) {
+	lock, err := i.opts.locker.GetLockWithID(ctx, req.VolumeId)
+	if err != nil {
+		return nil, err
+	}
+	if closer, ok := lock.(io.Closer); ok {
+		defer closer.Close() // #nosec G307
+	}
+	if !lock.TryLock(i.opts.timeout) {
+		return nil, status.Error(codes.Aborted, pending)
+	}
+	defer lock.Unlock()
+
+	return handler(ctx, req)
+}
+
+func (i *interceptor) createVolume(ctx context.Context, req *csi.CreateVolumeRequest,
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, resErr error) {
+
+	lock, err := i.opts.locker.GetLockWithID(ctx, req.Name)
+	if err != nil {
+		return nil, err
```

Comment:
please check this it should return the err only according to your condition , but you are returning the nil and err
any specific reason ?